src/spesh/arg_guard.c

```diff
@@ -383,12 +383,18 @@ void MVM_spesh_arg_guard_regenerate(MVMThreadContext *tc, MVMSpeshArgGuard **gua
 
     /* Install the produced argument guard. */
     if (*guard_ptr) {
-        MVMSpeshArgGuard *prev = *guard_ptr;
-        *guard_ptr = tree;
+        /* Ensure that the tree is already fully out of our cache.
+         * This is not necessary on x86_64, which has
+         * Total Store Ordering, but architectures like ARM are weaker
+         * and other cores can observe older contents of the  */
+        MVM_barrier();
+        MVMSpeshArgGuard *prev = (MVMSpeshArgGuard *)MVM_load(guard_ptr);
+        MVM_store(guard_ptr, tree);
         MVM_spesh_arg_guard_destroy(tc, prev, 1);
     }
     else {
-        *guard_ptr = tree;
+        MVM_barrier();
```

Comment:
Why is this barrier needed? I can see in the other branch when there's both a load and a store, but there's just a store here and guard_ptr and tree aren't accessed after.

Reply:
From IRC "timo: the stores that we are worried about are all the stores to the tree->nodes[i] further above".